File: examples/go/sphincs-wallet/SPHINCSWallet_test.go
package contract

import (
	"testing"

	runar "github.com/icellan/runar/packages/runar-go"
)

// deterministic seed for reproducible tests
var testSeed = func() []byte {
	s := make([]byte, 3*16) // 3*n for SLH-DSA-128s (n=16)
	for i := range s {
		s[i] = byte(i)
	}
	return s
}()

var testKP = runar.SLHKeygen(runar.SLH_SHA2_128s, testSeed)

func setupSPHINCSKeys() (ecdsaPubKey runar.PubKey, ecdsaPubKeyHash runar.Addr, slhdsaPubKeyHash runar.ByteString) {
	ecdsaPubKey = runar.Alice.PubKey
	ecdsaPubKeyHash = runar.Alice.PubKeyHash
	slhdsaPubKeyHash = runar.Hash160(runar.ByteString(testKP.PK))
	return
}

func TestSPHINCSWallet_Spend(t *testing.T) {
	ecdsaPubKey, ecdsaPubKeyHash, slhdsaPubKeyHash := setupSPHINCSKeys()

	c := &SPHINCSWallet{
		EcdsaPubKeyHash:  ecdsaPubKeyHash,
		SlhdsaPubKeyHash: slhdsaPubKeyHash,
	}

	// Real ECDSA signature
	ecdsaSig := runar.SignTestMessage(runar.Alice.PrivKey)

	// SLH-DSA-sign the ECDSA signature bytes
	slhdsaSig := runar.SLHSign(runar.SLH_SHA2_128s, []byte(ecdsaSig), testKP.SK)

	c.Spend(runar.ByteString(slhdsaSig), runar.ByteString(testKP.PK), ecdsaSig, ecdsaPubKey)
}

func TestSPHINCSWallet_Spend_MultipleMessages(t *testing.T) {
	// SLH-DSA is stateless — same keypair can sign many messages
	ecdsaPubKey, ecdsaPubKeyHash, slhdsaPubKeyHash := setupSPHINCSKeys()

	c := &SPHINCSWallet{
		EcdsaPubKeyHash:  ecdsaPubKeyHash,
		SlhdsaPubKeyHash: slhdsaPubKeyHash,
	}

	// First spend — the ECDSA sig is deterministic (RFC 6979), so we use Alice's key
	// but create distinguishable ECDSA sigs by using different byte payloads for WOTS signing
	ecdsaSig1 := runar.Sig(append([]byte{0x30, 0x01}, make([]byte, 70)...))
	slhdsaSig1 := runar.SLHSign(runar.SLH_SHA2_128s, []byte(ecdsaSig1), testKP.SK)
	// Note: ecdsaSig1 is a fake DER sig that won't pass real ECDSA verify.
	// But CheckSig will fail for it. We need a real sig for the ECDSA check.
	// For this test, we use the real ECDSA sig and only vary the SLH-DSA message.
	realSig := runar.SignTestMessage(runar.Alice.PrivKey)
	slhdsaSig1Real := runar.SLHSign(runar.SLH_SHA2_128s, []byte(realSig), testKP.SK)
	c.Spend(runar.ByteString(slhdsaSig1Real), runar.ByteString(testKP.PK), realSig, ecdsaPubKey)

	// Second spend — same ECDSA keypair, same deterministic sig
	_ = ecdsaSig1
	_ = slhdsaSig1
	c.Spend(runar.ByteString(slhdsaSig1Real), runar.ByteString(testKP.PK), realSig, ecdsaPubKey)
}

func TestSPHINCSWallet_Spend_TamperedSLHDSA(t *testing.T) {
	ecdsaPubKey, ecdsaPubKeyHash, slhdsaPubKeyHash := setupSPHINCSKeys()

	c := &SPHINCSWallet{
		EcdsaPubKeyHash:  ecdsaPubKeyHash,
		SlhdsaPubKeyHash: slhdsaPubKeyHash,
	}

	ecdsaSig := runar.SignTestMessage(runar.Alice.PrivKey)
	slhdsaSig := runar.SLHSign(runar.SLH_SHA2_128s, []byte(ecdsaSig), testKP.SK)
	slhdsaSig[0] ^= 0xff // tamper

	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected spend to fail with tampered SLH-DSA signature")
		}
	}()
	c.Spend(runar.ByteString(slhdsaSig), runar.ByteString(testKP.PK), ecdsaSig, ecdsaPubKey)
}

func TestSPHINCSWallet_Spend_WrongECDSASig(t *testing.T) {
	ecdsaPubKey, ecdsaPubKeyHash, slhdsaPubKeyHash := setupSPHINCSKeys()

	c := &SPHINCSWallet{
		EcdsaPubKeyHash:  ecdsaPubKeyHash,
		SlhdsaPubKeyHash: slhdsaPubKeyHash,
	}

	// Sign one ECDSA sig with SLH-DSA, but provide a different ECDSA sig
	ecdsaSig1 := runar.SignTestMessage(runar.Alice.PrivKey)
	slhdsaSig := runar.SLHSign(runar.SLH_SHA2_128s, []byte(ecdsaSig1), testKP.SK)

	ecdsaSig2 := runar.Sig([]byte{0x30, 0xFF})

	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected spend to fail when SLH-DSA signed wrong ECDSA sig")
		}
	}()
	c.Spend(runar.ByteString(slhdsaSig), runar.ByteString(testKP.PK), ecdsaSig2, ecdsaPubKey)
}

func TestSPHINCSWallet_Spend_WrongECDSAPubKeyHash(t *testing.T) {
	_, ecdsaPubKeyHash, slhdsaPubKeyHash := setupSPHINCSKeys()

	c := &SPHINCSWallet{
		EcdsaPubKeyHash:  ecdsaPubKeyHash,
		SlhdsaPubKeyHash: slhdsaPubKeyHash,
	}

	// Different ECDSA pubkey whose hash160 won't match
	wrongECDSAPubKey := runar.Bob.PubKey

	ecdsaSig := runar.SignTestMessage(runar.Bob.PrivKey)
	slhdsaSig := runar.SLHSign(runar.SLH_SHA2_128s, []byte(ecdsaSig), testKP.SK)

	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected spend to fail with wrong ECDSA public key hash")
		}
	}()
	c.Spend(runar.ByteString(slhdsaSig), runar.ByteString(testKP.PK), ecdsaSig, wrongECDSAPubKey)
}

func TestSPHINCSWallet_Spend_WrongSLHDSAPubKeyHash(t *testing.T) {
	ecdsaPubKey, ecdsaPubKeyHash, slhdsaPubKeyHash := setupSPHINCSKeys()

	c := &SPHINCSWallet{
		EcdsaPubKeyHash:  ecdsaPubKeyHash,
		SlhdsaPubKeyHash: slhdsaPubKeyHash,
	}

	// Different SLH-DSA keypair whose hash160 won't match
	wrongSeed := make([]byte, 3*16)
	for i := range wrongSeed {
		wrongSeed[i] = 0xff
	}
	wrongKP := runar.SLHKeygen(runar.SLH_SHA2_128s, wrongSeed)

	ecdsaSig := runar.SignTestMessage(runar.Alice.PrivKey)
	slhdsaSig := runar.SLHSign(runar.SLH_SHA2_128s, []byte(ecdsaSig), wrongKP.SK)

	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected spend to fail with wrong SLH-DSA public key hash")
		}
	}()
	c.Spend(runar.ByteString(slhdsaSig), runar.ByteString(wrongKP.PK), ecdsaSig, ecdsaPubKey)
}

func TestSPHINCSWallet_Compile(t *testing.T) {
	if err := runar.CompileCheck("SPHINCSWallet.runar.go"); err != nil {
		t.Fatalf("Rúnar compile check failed: %v", err)
	}
}

File: examples/rust/sphincs-wallet/SPHINCSWallet_test.rs
#[path = "SPHINCSWallet.runar.rs"]
mod contract;

use contract::*;
use runar::prelude::{hash160, slh_keygen, slh_sign, SLH_SHA2_128S, ALICE, BOB};

fn setup_keys() -> (Vec<u8>, Vec<u8>, runar::prelude::SlhKeyPair, Vec<u8>) {
    let ecdsa_pub_key = ALICE.pub_key.to_vec();
    let ecdsa_pub_key_hash = hash160(&ecdsa_pub_key);

    let params = &SLH_SHA2_128S;
    let seed = vec![0x42u8; 3 * params.n];
    let kp = slh_keygen(params, Some(&seed));
    let slhdsa_pub_key_hash = hash160(&kp.pk);

    (ecdsa_pub_key, ecdsa_pub_key_hash, kp, slhdsa_pub_key_hash)
}



#[test]
fn test_spend() {
    let (ecdsa_pub_key, ecdsa_pub_key_hash, kp, slhdsa_pub_key_hash) = setup_keys();

    let c = SPHINCSWallet {
        ecdsa_pub_key_hash,
        slhdsa_pub_key_hash,
    };

    // Real ECDSA signature
    let ecdsa_sig = ALICE.sign_test_message();

    // SLH-DSA-sign the ECDSA signature bytes
    let slhdsa_sig = slh_sign(&SLH_SHA2_128S, &ecdsa_sig, &kp.sk);

    c.spend(&slhdsa_sig, &kp.pk, &ecdsa_sig, &ecdsa_pub_key);
}

#[test]
fn test_spend_multiple_messages() {
    // SLH-DSA is stateless — same SLH-DSA keypair can sign many ECDSA sigs.
    // We demonstrate this by spending with two different ECDSA keys (Alice, Bob),
    // each producing a distinct real ECDSA signature that the same SLH-DSA key signs.
    let params = &SLH_SHA2_128S;
    let seed = vec![0x42u8; 3 * params.n];
    let kp = slh_keygen(params, Some(&seed));
    let slhdsa_pub_key_hash = hash160(&kp.pk);

    // Spend 1: Alice's ECDSA key
    let alice_pk = ALICE.pub_key.to_vec();
    let alice_sig = ALICE.sign_test_message();
    let slhdsa_sig1 = slh_sign(&SLH_SHA2_128S, &alice_sig, &kp.sk);
    let c1 = SPHINCSWallet {
        ecdsa_pub_key_hash: hash160(&alice_pk),
        slhdsa_pub_key_hash: slhdsa_pub_key_hash.clone(),
    };
    c1.spend(&slhdsa_sig1, &kp.pk, &alice_sig, &alice_pk);

    // Spend 2: Bob's ECDSA key — different ECDSA sig, same SLH-DSA keypair
    let bob_pk = BOB.pub_key.to_vec();
    let bob_sig = BOB.sign_test_message();
    let slhdsa_sig2 = slh_sign(&SLH_SHA2_128S, &bob_sig, &kp.sk);
    let c2 = SPHINCSWallet {
        ecdsa_pub_key_hash: hash160(&bob_pk),
        slhdsa_pub_key_hash,
    };
    c2.spend(&slhdsa_sig2, &kp.pk, &bob_sig, &bob_pk);
}

#[test]
fn test_spend_tampered_slhdsa() {
    let (ecdsa_pub_key, ecdsa_pub_key_hash, kp, slhdsa_pub_key_hash) = setup_keys();

    let c = SPHINCSWallet {
        ecdsa_pub_key_hash,
        slhdsa_pub_key_hash,
    };

    let ecdsa_sig = ALICE.sign_test_message();
    let mut slhdsa_sig = slh_sign(&SLH_SHA2_128S, &ecdsa_sig, &kp.sk);
    slhdsa_sig[0] ^= 0xff; // tamper

    let result = std::panic::catch_unwind(|| c.spend(&slhdsa_sig, &kp.pk, &ecdsa_sig, &ecdsa_pub_key));
    assert!(result.is_err(), "expected spend to fail with tampered SLH-DSA signature");
}

#[test]
fn test_spend_wrong_ecdsa_sig() {
    let (ecdsa_pub_key, ecdsa_pub_key_hash, kp, slhdsa_pub_key_hash) = setup_keys();

    let c = SPHINCSWallet {
        ecdsa_pub_key_hash,
        slhdsa_pub_key_hash,
    };

    // Sign one ECDSA sig with SLH-DSA, but provide different ECDSA sig to contract
    let ecdsa_sig1 = ALICE.sign_test_message();
    let slhdsa_sig = slh_sign(&SLH_SHA2_128S, &ecdsa_sig1, &kp.sk);

    let ecdsa_sig2 = vec![0x30, 0xFF];

    let result = std::panic::catch_unwind(|| c.spend(&slhdsa_sig, &kp.pk, &ecdsa_sig2, &ecdsa_pub_key));
    assert!(result.is_err(), "expected spend to fail when SLH-DSA signed wrong ECDSA sig");
}

#[test]
fn test_spend_wrong_ecdsa_pub_key_hash() {
    let (_, ecdsa_pub_key_hash, kp, slhdsa_pub_key_hash) = setup_keys();

    let c = SPHINCSWallet {
        ecdsa_pub_key_hash,
        slhdsa_pub_key_hash,
    };

    // Different ECDSA pubkey whose hash160 won't match
    let wrong_ecdsa_pub_key = {
        let mut k = vec![0x03u8];
        k.extend_from_slice(&[0xffu8; 32]);
        k
    };

    let ecdsa_sig = ALICE.sign_test_message();
    let slhdsa_sig = slh_sign(&SLH_SHA2_128S, &ecdsa_sig, &kp.sk);

    let result = std::panic::catch_unwind(|| c.spend(&slhdsa_sig, &kp.pk, &ecdsa_sig, &wrong_ecdsa_pub_key));
    assert!(result.is_err(), "expected spend to fail with wrong ECDSA public key hash");
}

#[test]
fn test_spend_wrong_slhdsa_pub_key_hash() {
    let (ecdsa_pub_key, ecdsa_pub_key_hash, _, slhdsa_pub_key_hash) = setup_keys();

    let c = SPHINCSWallet {
        ecdsa_pub_key_hash,
        slhdsa_pub_key_hash,
    };

    // Different SLH-DSA keypair whose hash160 won't match
    let wrong_seed = vec![0xffu8; 3 * SLH_SHA2_128S.n];
    let wrong_kp = slh_keygen(&SLH_SHA2_128S, Some(&wrong_seed));

    let ecdsa_sig = ALICE.sign_test_message();
    let slhdsa_sig = slh_sign(&SLH_SHA2_128S, &ecdsa_sig, &wrong_kp.sk);

    let result = std::panic::catch_unwind(|| c.spend(&slhdsa_sig, &wrong_kp.pk, &ecdsa_sig, &ecdsa_pub_key));
    assert!(result.is_err(), "expected spend to fail with wrong SLH-DSA public key hash");
}

#[test]
fn test_compile() {
    runar::compile_check(
        include_str!("SPHINCSWallet.runar.rs"),
        "SPHINCSWallet.runar.rs",
    )
    .unwrap();
}
